### PR TITLE
Add of iterOptThreshold segmentation Algorithm

### LINKED ISSUE #1: Image Segmentation
It is **required** to add 1-3 slides with samples to the presentation according to the Subj.
Each slide shall contain pictures with short description and link to the source if exists.
**Pull request** shall be created and closed before milestone end.

### LINKED ISSUE #18: Iterative Optimal Thresholding
## Task

It is required to implement Demo application according to the Subj.
## Requirements

Demo application shall show how algorithm work step by step. See sample of the external [Region Growing Demo](https://www.youtube.com/watch?v=IjK774GwRGk).
- [x] input image shall be loaded from the file specified in argv parameter.
- [x] command line parser shall be used for parsing argc, argv parameters.
- [x] If user input is required to specify seeds, starting points, etc. then UI shall handle mouse events.
- [x] keyboard events shall be handled
  - [x] to restore (R key) demonstration.
  - [x] launch (Space key) demonstration.
  - [x] next step (N key).
  - [x] exit (ESC key) demonstration.
- [x] print algorithm's execution time.
- [x] dump initial image, image with user enter, all algorithm steps and overlaid colored segmentation over initial image: {algname_imgname_##.png}
- [x] in thresholding methods separate window with histogram shall be shown
- [x] histogram image also shall be dump in image file step by step.

Any question shall be discussed in comment to this issue or in group chat in VK.
## Pull request
- [x] Pull request shall contain algorithm description and screenshots of the intermediate pictures demonstrated how algorithm work.
### Checklist: Pull request shall:
- [ ] be assigned on reviewer which may be found from the group list = (Your# + Lesson# - 1) mod 8.
- [x] have label according to the lesson.
- [x] milestone according to the lesson.
- [x] contain text (e.g. closes #1) in title which [trigger issue closing](https://help.github.com/articles/closing-issues-via-commit-messages/).

It is responsible of the student to push reviewer. When reviewer approves merge then "Pull Request" shall be assigned to the teacher (@rgolovanov).

**Application shall be implemented and code shall be reviewed before iteration end.**

### PR DESCRIPTION
**Pull request by NickLyrick**
# Lesson 2. Iterative Optimal Threashold segmentation. Close #18

Algorithm:
1. Choose an initial value of the T.
2. Segment the image with T which produce two groups: G_1 contains pixels with intensity ≤T and G_2 contains pixels with intensity >T.
3. Calculating avarage intensivity μ1,μ2 for G1,G2 respectively.
4. Calculating new threshold: T=(μ1+μ2)/2.
5. Repeat 2-5 untill differense of the T for the last and previus iteration is less then some δ.
# Results

First hist
iteroptthreshold_0_hist
First step
iteroptthreshold_frame_0

Third hist
iteroptthreshold_3_hist
Third step
iteroptthreshold_frame_3

Last hist
iteroptthreshold_7_hist
Last step
iteroptthreshold_frame_7
# Task

It is required to implement Demo application according to the Subj.
# Requirements

Demo application shall show how algorithm work step by step. See sample of the external Region Growing Demo.
- [x]  input image shall be loaded from the file specified in argv parameter. 
- [x] command line parser shall be used for parsing argc, argv parameters.
- [x]  keyboard events shall be handled
- [x]  If user input is required to specify seeds, starting points, etc. then UI shall handle mouse events.
  - [x]  to restore (R key) demonstration.
  - [x]  launch (Space key) demonstration.
  - [x]  next step (N key).
  - [x]  exit (ESC key) demonstration.
  - [x]  print algorithm's execution time.
- [x]  dump initial image, image with user enter, all algorithm steps and overlaid colored segmentation over
      initial image: {algname_imgname_##.png}
- [x]  in thresholding methods separate window with histogram shall be shown
- [x]  histogram image also shall be dump in image file step by step.

Any question shall be discussed in comment to this issue or in group chat in VK.
# Pull request
- [x]  Pull request shall contain algorithm description and screenshots of the intermediate pictures 
  demonstrated how algorithm work.
# Checklist: Pull request shall:
- [x]  be assigned on reviewer which may be found from the group list = (Your# + Lesson# - 1) mod 8.
- [x]  have label according to the lesson.
- [x]  milestone according to the lesson.
- [x]  contain text (e.g. closes #1) in title which trigger issue closing.

It is responsible of the student to push reviewer. When reviewer approves merge then "Pull Request" shall be assigned to the teacher (@rgolovanov).

Application shall be implemented and code shall be reviewed before iteration end.

### CODE REVIEW

File: lesson2/main.cpp [hunk 1]
```diff
@@ -40,15 +40,15 @@ class DemoRegistry {
 };
 
 // Declaration of Demo functions implemented in separate unit
-int iterOptThreshold(int, char**) { return 0; /*TODO: implement in separate file*/ }
-int otsu(int, char**) { return 0; /*TODO: implement in separate file*/ }
-int gmm(int, char**) { return 0; /*TODO: implement in separate file*/ }
-int regionGrowing(int, char**) { return 0; /*TODO: implement in separate file*/ }
-int splitAndMerge(int, char**) { return 0; /*TODO: implement in separate file*/ }
+int iterOptThreshold(int, char**);	/*TODO: implement in separate file*/
```
Comment (rgolovanov) [inline lesson2/main.cpp]: нельзя изменять строки, относящиеся к другим алгоритмам. Теперь замежить не получится из-за конфликтов

**Review by rgolovanov — COMMENTED** [review]
Изображение не очень показательно для демонстрации алгоритма сегментации из-за рамытых границ и бликов. Вместо этого изображения используй какое-нибудь классическое, например из набора демо изображений Matlab-а

File: lesson2/main.cpp (hunk 1 above)
Comment (NickLyrick) [reply lesson2/main.cpp]: @rgolovanov, 
1. изменил картинку на rice.png из примеров MATLAB. 
2. Вернул main в исходное состояние, которое склонировал.

**Comment by rgolovanov** [comment]
@NickLyrick попрежнему конфликт с текущей версией ветки master. попробуй у себя локально сделать **git pull --rebase**, а затем устранить конфликты.

**Pull request merged**